From f5ed1a21e86659c98b53c067fb5c690487ad89d1 Mon Sep 17 00:00:00 2001
From: Peng Fan <peng.fan@nxp.com>
Date: Wed, 13 Jan 2021 11:39:46 +0100
Subject: [PATCH 048/184] virtio: ivshmem: check peer_state early

We enable the virtio_ivshmem driver and add
shared memory region including pci devices in cell file. But we not
start backend.

There might be garbage data in "vi_dev->virtio_header", so we need to
check peer_state to abort the probe earlier.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
---
 drivers/virtio/virtio_ivshmem.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/virtio/virtio_ivshmem.c b/drivers/virtio/virtio_ivshmem.c
index bf6237779..7132cf0e2 100644
--- a/drivers/virtio/virtio_ivshmem.c
+++ b/drivers/virtio/virtio_ivshmem.c
@@ -846,6 +846,10 @@ static int virtio_ivshmem_probe(struct pci_dev *pci_dev,
 		return -ENOMEM;
 
 	vi_dev->peer_state = &state_table[vi_dev->peer_id];
+	if (*vi_dev->peer_state != VIRTIO_STATE_READY) {
+		dev_err(&pci_dev->dev, "backend not ready\n");
+		return -ENODEV;
+	}
 
 	section_addr += section_sz;
 
@@ -898,11 +902,6 @@ static int virtio_ivshmem_probe(struct pci_dev *pci_dev,
 
 	set_dma_ops(&pci_dev->dev, &virtio_ivshmem_dma_ops);
 
-	if (*vi_dev->peer_state != VIRTIO_STATE_READY) {
-		dev_err(&pci_dev->dev, "backend not ready\n");
-		return -ENODEV;
-	}
-
 	pci_set_master(pci_dev);
 	pci_write_config_byte(pci_dev, vendor_cap + IVSHM_CFG_PRIV_CNTL, 0);
 
-- 
2.39.2

